File: modules/zig/src/main/java/com/falsepattern/zigbrains/zig/formatter/ZigBlock.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.zig.formatter;

import com.intellij.formatting.Alignment;
import com.intellij.formatting.Block;
import com.intellij.formatting.Indent;
import com.intellij.formatting.Spacing;
import com.intellij.formatting.SpacingBuilder;
import com.intellij.formatting.Wrap;
import com.intellij.lang.ASTNode;
import com.intellij.psi.TokenType;
import com.intellij.psi.formatter.common.AbstractBlock;
import com.intellij.psi.tree.IElementType;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import static com.falsepattern.zigbrains.zig.psi.ZigTypes.ADDITION_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.ASM_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.ASM_INPUT_LIST;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.ASM_OUTPUT_LIST;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.BITWISE_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.BIT_SHIFT_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.BLOCK;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.BOOL_AND_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.BOOL_OR_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.COMPARE_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.CONTAINER_DECL_AUTO;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.CONTAINER_DECL_TYPE;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.CONTAINER_DOC_COMMENT;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.EXPR_LIST;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.FN_CALL_ARGUMENTS;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.FN_PROTO;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.IF_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.IF_PREFIX;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.IF_STATEMENT;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.INIT_LIST;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.KEYWORD_ASM;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.KEYWORD_ELSE;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.KEYWORD_VOLATILE;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.LBRACE;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.LPAREN;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.MULTIPLY_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.PARAM_DECL_LIST;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.RBRACE;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.RPAREN;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.SWITCH_EXPR;
import static com.falsepattern.zigbrains.zig.psi.ZigTypes.SWITCH_PRONG_LIST;

public class ZigBlock extends AbstractBlock {

    private final SpacingBuilder spacingBuilder;

    protected ZigBlock(@NotNull ASTNode node, @Nullable Wrap wrap, @Nullable Alignment alignment, SpacingBuilder spacingBuilder) {
        super(node, wrap, alignment);
        this.spacingBuilder = spacingBuilder;
    }


    @Override
    protected List<Block> buildChildren() {
        var blocks = new ArrayList<Block>();
        var child = myNode.getFirstChildNode();
        while (child != null) {
            if (child.getElementType() != TokenType.WHITE_SPACE) {
                var block = new ZigBlock(child, null, null, spacingBuilder);
                blocks.add(block);
            }
            child = child.getTreeNext();
        }

        return blocks;
    }

    @Override
    public @Nullable Spacing getSpacing(@Nullable Block block, @NotNull Block block1) {
        return null;
    }

    @Override
    public boolean isLeaf() {
        return myNode.getFirstChildNode() == null;
    }

    private static final IElementType PLACEHOLDER = new IElementType("placeholder", null);

    @Override
    protected @Nullable Indent getChildIndent() {
        val node = getNode();
        return getIndentBasedOnParentType(node, null, node.getElementType(), PLACEHOLDER);
    }

    @Override
    public Indent getIndent() {
        val node = getNode();
        val parent = node.getTreeParent();
        if (parent != null) {
            return getIndentBasedOnParentType(parent, node, parent.getElementType(), node.getElementType());
        }
        return Indent.getNoneIndent();
    }

    private static boolean isBrace(IElementType element) {
        return element == LBRACE || element == RBRACE;
    }

    private static boolean isParen(IElementType element) {
        return element == LPAREN || element == RPAREN;
    }

    private static boolean isEmpty(ASTNode child) {
        return child.getFirstChildNode() == null;
    }


    private static @NotNull Indent getIndentBasedOnParentType(@NotNull ASTNode parent, @Nullable ASTNode child, @NotNull IElementType parentElementType, @NotNull IElementType childElementType) {
        //Statement blocks
        if (parentElementType == BLOCK && !isBrace(childElementType))
            return Indent.getNormalIndent();

        //Struct/tuple initializers
        if (parentElementType == INIT_LIST && !isBrace(childElementType))
            return Indent.getNormalIndent();

        //Function call args
        if (parentElementType == EXPR_LIST ||
            parentElementType == FN_CALL_ARGUMENTS && childElementType == PLACEHOLDER)
            return Indent.getNormalIndent();

        //Function declaration parameters
        if (parentElementType == PARAM_DECL_LIST ||
            parentElementType == FN_PROTO && childElementType == PLACEHOLDER)
            return Indent.getNormalIndent();

        //Chained operations on newlines
        if ((parentElementType == BOOL_OR_EXPR ||
            parentElementType == BOOL_AND_EXPR ||
            parentElementType == COMPARE_EXPR ||
            parentElementType == BITWISE_EXPR ||
            parentElementType == BIT_SHIFT_EXPR ||
            parentElementType == ADDITION_EXPR ||
            parentElementType == MULTIPLY_EXPR) &&
            parent.getFirstChildNode() != child)
            return Indent.getNormalIndent();

        //Switch prongs
        if (parentElementType == SWITCH_PRONG_LIST ||
            parentElementType == SWITCH_EXPR && childElementType == PLACEHOLDER)
            return Indent.getNormalIndent();

        //If expressions/statements
        if ((parentElementType == IF_EXPR || parentElementType == IF_STATEMENT) && childElementType != KEYWORD_ELSE && childElementType != IF_PREFIX)
            return Indent.getNormalIndent();

        //Struct members
        if (parentElementType == CONTAINER_DECL_AUTO && childElementType != CONTAINER_DECL_TYPE && childElementType != CONTAINER_DOC_COMMENT && !isBrace(childElementType))
            return Indent.getNormalIndent();

        //Inline assembly body
        if (parentElementType == ASM_EXPR && childElementType != KEYWORD_ASM && childElementType != KEYWORD_VOLATILE && !isParen(childElementType))
            return Indent.getNormalIndent();

        //Assembly params
        if (parentElementType == ASM_INPUT_LIST || parentElementType == ASM_OUTPUT_LIST)
            return Indent.getSpaceIndent(2);

        return Indent.getNoneIndent();
    }
}

File: modules/core/src/main/kotlin/com/falsepattern/zigbrains/zig/formatter/ZigBlock.kt
package com.falsepattern.zigbrains.zig.formatter

import com.intellij.formatting.*
import com.intellij.lang.ASTNode
import com.intellij.psi.TokenType
import com.intellij.psi.formatter.common.AbstractBlock
import com.intellij.psi.tree.IElementType
import com.falsepattern.zigbrains.zig.psi.ZigTypes.*

class ZigBlock(
    node: ASTNode,
    wrap: Wrap?,
    alignment: Alignment?,
    val spacingBuilder: SpacingBuilder
) : AbstractBlock(node, wrap, alignment) {

    override fun getSpacing(child1: Block?, child2: Block) = null

    override fun isLeaf() = myNode.firstChildNode == null

    override fun buildChildren(): MutableList<Block> {
        val blocks = ArrayList<Block>()
        var child = myNode.firstChildNode
        while (child != null) {
            if (child.elementType != TokenType.WHITE_SPACE) {
                blocks.add(ZigBlock(child, null, null, spacingBuilder))
            }
            child = child.treeNext
        }

        return blocks
    }

    override fun getChildIndent(): Indent {
        val node = this.node
        return getIndentBasedOnParentType(node, null, node.elementType, PLACEHOLDER)
    }

    override fun getIndent(): Indent {
        val node = this.node
        val parent = node.treeParent ?: return noneIndent
        return getIndentBasedOnParentType(parent, node, parent.elementType, node.elementType)
    }
}

private fun getIndentBasedOnParentType(
    parent: ASTNode,
    child: ASTNode?,
    parentType: IElementType,
    childType: IElementType
): Indent {
    //Statement blocks
    if (parentType == BLOCK && !childType.isBrace)
        return normalIndent

    //Struct/tuple initializers
    if (parentType == INIT_LIST && !childType.isBrace)
        return normalIndent

    //Function call args
    if (parentType == EXPR_LIST ||
        parentType == FN_CALL_ARGUMENTS && childType === PLACEHOLDER
    )
        return normalIndent

    //Function declaration parameters
    if (parentType == PARAM_DECL_LIST ||
        parentType == FN_PROTO && childType === PLACEHOLDER
    )
        return normalIndent

    //Chained operations on newlines
    if ((parentType == BOOL_OR_EXPR ||
         parentType == BOOL_AND_EXPR ||
         parentType == COMPARE_EXPR ||
         parentType == BITWISE_EXPR ||
         parentType == BIT_SHIFT_EXPR ||
         parentType == ADDITION_EXPR ||
         parentType == MULTIPLY_EXPR) &&
        parent.firstChildNode != child
    )
        return normalIndent

    //Switch prongs
    if (parentType == SWITCH_PRONG_LIST ||
        parentType == SWITCH_EXPR && childType == PLACEHOLDER
    )
        return normalIndent

    //If expressions/statements
    if ((parentType == IF_EXPR ||
         parentType == IF_STATEMENT) &&
        childType != KEYWORD_ELSE &&
        childType != IF_PREFIX
    )
        return normalIndent

    //Struct members
    if (parentType == CONTAINER_DECL_AUTO &&
        childType != CONTAINER_DECL_TYPE &&
        childType != CONTAINER_DOC_COMMENT &&
        !childType.isBrace
    )
        return normalIndent

    //Inline assembly body
    if (parentType == ASM_EXPR &&
        childType != KEYWORD_ASM &&
        childType != KEYWORD_VOLATILE &&
        !childType.isParen
    )
        return normalIndent

    //Assembly params
    if (parentType == ASM_INPUT_LIST || parentType == ASM_OUTPUT_LIST)
        return spaceIndent(2)

    //Variable declarations
    if (parentType == VAR_DECL_EXPR_STATEMENT &&
        (childType == PLACEHOLDER ||
         child?.treePrevNonSpace?.elementType == EQUAL)
    )
        return normalIndent

    return noneIndent
}

private val ASTNode.treePrevNonSpace: ASTNode? get() {
    var it = this.treePrev
    while (it?.elementType == TokenType.WHITE_SPACE) {
        it = it.treePrev
    }
    return it
}

private val normalIndent: Indent get() = Indent.getNormalIndent()

private fun spaceIndent(spaces: Int) = Indent.getSpaceIndent(2)

private val noneIndent: Indent get() = Indent.getNoneIndent()

private val IElementType?.isBrace: Boolean
    get() {
        return this == LBRACE || this == RBRACE
    }

private val IElementType?.isParen: Boolean
    get() {
        return this == LPAREN || this == RPAREN
    }

private val PLACEHOLDER = IElementType("placeholder", null)